Severity-based logging -- parent PR

jonatack opened the pull request:
This is a parent PR for continuing updates to severity-based logging. See https://github.com/bitcoin/bitcoin/issues/20576 for motivation and https://github.com/bitcoin/bitcoin/pull/25306 for discussion.

- [x] https://github.com/bitcoin/bitcoin/pull/24464
- [x] https://github.com/bitcoin/bitcoin/pull/25286
- [x] #25306
- [x] https://github.com/bitcoin/bitcoin/issues/25614
- [x] https://github.com/bitcoin/bitcoin/pull/26957
- [x] https://github.com/bitcoin/bitcoin/pull/27632
- #27231 
- print the wallet name more clearly in wallet logging to distinguish it from category/level
- unconditionally log `Info` severity level messages
- replace the hardcoded `LogLevelsList()` vector with a programmatic one derived from the `Level` enum class
- convert `GetLogCategory()` to `std::optional`
- deduplicate the `LogCategory` code
- update to severity-based logging in the following areas, dropping the use of `LogPrintf` and `LogPrint`: addrdb, addrman, banman, i2p, mempool, netbase, net, net_processing, timedata, torcontrol
- when we're ready to drop the other logging macros, rename `LogPrintLevel` to simply `Log`, which will be the only log macro needed after migration from the other macros is complete

laanwj commented on src/i2p.cpp:
```diff
@@ -376,7 +376,7 @@ void Session::CreateIfNotCreatedAlready()
     m_control_sock = std::move(sock);
 
-    LogPrintf("I2P: SAM session created: session id=%s, my address=%s\n", m_session_id,
+    LogPrintf("[i2p] SAM session created: session id=%s, my address=%s\n", m_session_id,
```
I understand the idea, but I don't think we should do it this way by inserting `[cat]` manually. We need a way to log things with a category even without `-debug=`.
This is why I have a commit in #25202 to log everything >= Warning unconditionally. I'm not sure that is the best solution but I like it more than this.

jonatack replied:
Thanks, SGTM, will just do the de-dupe then (also for ZMQ and PRUNE).

laanwj replied:
That's fine but to be clear I mean I agree conceptually with doing it, just not in this way.

jonatack replied:
Done.

jonatack replied:
> That's fine but to be clear I mean I agree conceptually with doing it, just not in this way.

Agree, thanks for clarifying.  I'll start with reviewing your change and then go from there / update here.

jonatack replied:
How about `LogPrintfCategory`?
```diff
--- a/src/logging.h
+++ b/src/logging.h
@@ -198,6 +198,7 @@ static inline void LogPrintf_(const std::string& logging_function, const std::st
 
 #define LogPrintf(...) LogPrintLevel_(BCLog::LogFlags::NONE, BCLog::Level::None, __VA_ARGS__)
+#define LogPrintfCategory(category, ...) LogPrintLevel_(category, BCLog::Level::None, __VA_ARGS__)

--- a/src/i2p.cpp
+++ b/src/i2p.cpp
@@ -376,7 +376,7 @@ void Session::CreateIfNotCreatedAlready()
 
-    LogPrintf("I2P: SAM session created: session id=%s, my address=%s\n", m_session_id,
+    LogPrintfCategory(BCLog::I2P, "SAM session created: session id=%s, my address=%s\n", m_session_id,
```

debug log
```
2022-05-25T16:03:02Z [i2p] SAM session created
```

jonatack replied:
Done

laanwj replied:
Yes, lgtm!

jonatack committed:
tor: log messages from torcontrol with category and debug level

jonatack committed:
i2p: log messages from I2P with category and debug level

jonatack committed:
refactor: remove unused i2p::Session::Log() template method

jonatack committed:
netbase: log netbase messages with category and debug level

and change a few of the NET category logging messages to PROXY when the message
relates to SOCKS5, per the following log category descriptions in
https://github.com/jonatack/bitcoin-development/blob/master/logging.md
which is an updated version of this Bitcoin Stack Exchange answer in 2017 by Andrew Chow:
https://bitcoin.stackexchange.com/questions/66892/what-are-the-debug-categories/66895

- net: All messages related to communicating with other nodes on the network,
  including what P2P messages were sent and received and to whom and other
  information about the network messages.

- proxy: Messages about using a SOCKS5 proxy and its authentication.

Before this commit, BCLog::PROXY was essentially unused (only one log message).

and reduce some unconditional logging that could be peer-provoked.

While here, improve the clarity of a few of the touched messages and use
consistent capitalization and naming, e.g. "Socks5()" vs "SOCKS5"

jonatack commented:
This will need to be updated after the merge of #25202 for the change in [refactor: Change LogPrintLevel order to category, severity](https://github.com/bitcoin/bitcoin/pull/25202/commits/c4e77177276ea2b79c4675cb2678ee2cc757b743), so leaving in draft for now.

jonatack committed:
net_processing: replace LogPrintf messages with category and debug level

using LogPrintLevel(), thereby removing all LogPrintf messages in this file.

Reduce some of the unconditional logging that could be peer-provoked
(disconnection for old chain, stalling, timeout, but not user-selected
cases like manual/noban/forcerelay peers), by moving it to the Debug level.

While here, improve the clarity of a few of the touched messages and use
consistent capitalization.

MarcoFalke commented on src/logging.h:
```diff
@@ -203,6 +203,7 @@ static inline void LogPrintf_(const std::string& logging_function, const std::st
 #define LogPrintLevel_(category, level, ...) LogPrintf_(__func__, __FILE__, __LINE__, category, level, __VA_ARGS__)
 
 #define LogPrintf(...) LogPrintLevel_(BCLog::LogFlags::NONE, BCLog::Level::None, __VA_ARGS__)
+#define LogPrintfCategory(category, ...) LogPrintLevel_(category, BCLog::Level::None, __VA_ARGS__)
```
Would be good to add a docstring for this. Is this meant to unconditionally log regardless of log level? If yes, it doesn't seem it would work. `BCLog::Level::None` is `1` but the check for unconditional logging is `>= Warning (3)`?

jonatack replied:
Good catch; this is wrong for `LogPrintf` as well. Likely going to fix by reordering the `BCLog::Level` enums:
```diff
diff --git a/src/logging.cpp b/src/logging.cpp
index f1a86f0dce..3f316f0101 100644
--- a/src/logging.cpp
+++ b/src/logging.cpp
@@ -187,8 +187,6 @@ bool GetLogCategory(BCLog::LogFlags& flag, const std::string& str)
 std::string LogLevelToStr(BCLog::Level level)
 {
     switch (level) {
-    case BCLog::Level::None:
-        return "none";
     case BCLog::Level::Debug:
         return "debug";
     case BCLog::Level::Info:
@@ -197,6 +195,8 @@ std::string LogLevelToStr(BCLog::Level level)
         return "warning";
     case BCLog::Level::Error:
         return "error";
+    case BCLog::Level::None:
+        return "";
     }
     assert(false);
 }
```
```diff
diff --git a/src/logging.h b/src/logging.h
index 8a896b6b33..f8e55b88b2 100644
--- a/src/logging.h
+++ b/src/logging.h
@@ -69,10 +69,10 @@ namespace BCLog {
     };
     enum class Level {
         Debug = 0,
-        None = 1,
-        Info = 2,
-        Warning = 3,
-        Error = 4,
+        Info,
+        Warning,
+        Error,
+        None,
     };
```

jonatack replied:
>  it doesn't seem it would work. BCLog::Level::None is 1 but the check for unconditional logging is >= Warning (3)?

Ah, `LogAcceptCategory()` with the logic you mention is called by neither `LogPrintf` nor `LogPrintfCategory`, so it seems fine as-is. Will add a docstring.

jonatack replied:
> Would be good to add a docstring for this.

Thanks for the suggestion; done.

jonatack committed:
net_processing: use severity-based logging for LogPrint messages

- Info for the minimum viable default logging for users

- Debug for high-level peer event messages: connection, disconnection,
  misbehavior, eviction

- Trace for logging of low-level, very high-frequency p2p messages.

While here, improve the clarity of the touched messages and use consistent
capitalization.

jonatack committed:
logging: simplify BCLog::Level enum class and LogLevelToStr() function

- simplify the BCLog::Level enum class (and future changes to it) by
  only setting the value of the first enumerator

- move the BCLog::Level:None enumerator to the end of the BCLog::Level
  enum class and LogLevelToStr() member function, as the None enumerator
  is only used internally, and by being the highest BCLog::Level value it
  can be used to iterate over the enumerators

- replace the unused BCLog::Level:None string "none" with an empty string
  in LogLevelToStr(); the case will currently never be hit

- add documentation

jonatack committed:
net, timedata: use severity-based logging for LogPrint messages

- Info for the minimum viable default logging for users

- Debug for high-level peer events: connection, disconnection,
  misbehavior, eviction

- Trace for logging of low-level, very high-frequency p2p messages

While here, improve the clarity of the touched messages and use consistent
capitalization.

jonatack committed:
logging, test: add BCLog::Level::Trace log severity level

for verbose log messages for development or debugging only, as bitcoind may run
more slowly, that are more granular/frequent than the Debug log level, i.e. for
very high-frequency, low-level messages to be logged distinctly from
higher-level, less-frequent debug logging that could still be usable in production.

An example would be to log higher-level peer events (connection, disconnection,
misbehavior, eviction) versus low-level, high-volume p2p messages in the
BCLog::NET category. This will enable the user to log only the former without
the latter, in order to focus on high-level peer management events.

With respect to the name, "trace" is suggested as the most granular level
in resources like the following:
- https://sematext.com/blog/logging-levels
- https://howtodoinjava.com/log4j2/logging-levels

Update the test framework and add test coverage.

jonatack committed:
net: replace LogPrintf messages with category and debug level

and reduce some of the unconditional logging that could be peer-provoked by
moving them to the Info and Debug levels:

- Info for the minimum viable default logging for users

- Debug for more detailed peer management messages: connection, disconnection,
  misbehavior, eviction, etc.

While here, improve the clarity of the touched messages and use consistent
capitalization.

klementtan committed:
refactor: deduplicate LogCategory code

laanwj commented on src/logging.h:
```diff
@@ -199,11 +200,14 @@ static inline void LogPrintf_(const std::string& logging_function, const std::st
 #define LogPrintf(...) LogPrintLevel_(BCLog::LogFlags::NONE, BCLog::Level::None, __VA_ARGS__)
 
+// Log unconditionally, prefixing the output with the passed category name.
+#define LogPrintfCategory(category, ...) LogPrintLevel_(category, BCLog::Level::None, __VA_ARGS__)
```
It's `LogPrintLevel` and `LogPrintfCategory` that's a bit inconsistent. I think we can leave out the `f` here.

klementtan commented:
Approach ACK

laanwj commented:
Concept ACK

klementtan committed:
logging: add -loglevel configuration option

- add a -loglevel=<level>|<category:level> config option, to allow users
  to set a global -loglevel and category-specific log levels

- do not print log messages having lower severity level than -loglevel

- add unit tests

jonatack commented:
Rebased/updated following the merge of #25286 that contained the first commit of this PR.

klementtan committed:
doc: release note for -loglevel configuration option

dergoegge commented on src/logging.h:
```diff
@@ -68,11 +68,12 @@ namespace BCLog {
-        Info = 2,
-        Warning = 3,
-        Error = 4,
+        Trace = 0,
```
I think some docs on the different levels wouldn't hurt. For example, it's not quite clear to me how you choose between `Trace` and `Debug`.

laanwj replied:
Generally, the distinction is:
- debug is for reasonably noisy logging, but not high enough volume to make it unusable in a production scenario
- tacing is super high detail, step by step, and will slow down the application and you'd definitely never want to enable it for a category unless you're developing or trouble-shooting a specific issue

jonatack replied:
Thanks! Added documentation to the Level enum class.

jonatack committed:
refactor: use std::optional for GetLogCategory()

jonatack committed:
logging: unconditionally log Info severity level messages

in addition to Warning and Error ones.

jonatack committed:
refactor: replace hardcoded LogLevelsList vector with a programmatic one

built from the BCLog:Level enum class, as done in GetNetworkNames().

jonatack commented:
Rebased and updated significantly to pull in the changes in #25287 (with review feedback and some clean-up and moving the refactoring commits to after the functional ones) and the changes discussed in #25306 (unconditionally log Info level and up). Will re-update the specific log messages themselves after the first commits here are confirmed or merged.

jonatack committed:
logging: replace LogPrintf ADDRMAN messages with category and debug level

jonatack committed:
logging: use severity-based logging for LogPrint ADDRMAN messages

jonatack committed:
logging: replace MEMPOOL LogPrint messages with category and debug level

jonatack committed:
logging: print wallet name more clearly to distinguish from category/level

jonatack committed:
doc: update logging in fuzzing.md Honggfuzz code

jonatack committed:
logging: replace mempool LogPrintf messages with category and debug level

jonatack commented:
Rebased and updated with the latest changes in subset PR #25614.

w0xlt reviewed (commented):
Approach ACK.

This PR improves significantly the quality of the log and the Log codebase.
I wonder if the main Log classes could be a separate library in another repository.
Awaiting rebase for ACK.

vasild commented:
Approach ACK

jonatack commented:
Up for grabs.

MarcoFalke commented:
Concept ACK

jonatack commented:
A big thank you to @vasild who rebased this branch to current master. Re-opening and updating.

dergoegge commented on src/net_processing.cpp:
```diff
@@ -2356,9 +2356,9 @@ void PeerManagerImpl::HeadersDirectFetchBlocks(CNode& pfrom, const Peer& peer, c
-            LogPrint(BCLog::NET, "Large reorg, won't direct fetch to %s (%d)\n",
-                    pindexLast->GetBlockHash().ToString(),
-                    pindexLast->nHeight);
+            LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "Large reorg, won't direct fetch to %s (%d)\n",
```
Is the switch to unconditional logging intended here?

If yes, then it's probably better done in a separate commit because it is not immediately clear that this is safe and it's quite easy to miss mixed in with all the other changes in this commit.

jonatack replied:
Yes, good point that any changes from conditional to unconditional logging should be separated out.

dergoegge commented on src/logging.cpp:
```diff
@@ -125,9 +125,9 @@ bool BCLog::Logger::WillLogCategory(BCLog::LogFlags category) const
-    if (level >= BCLog::Level::Warning) return true;
+    // Log the Info, Warning, and Error message levels unconditionally, so that
+    // important troubleshooting information isn't lost.
+    if (level >= BCLog::Level::Info) return true;
```
I feel like this almost deserves its own PR (maybe you are already planning on doing so). afaict we don't log anything right now using `Info` which is the only reason this change is safe but that might not be clear to reviewers of this rather large PR. (If for example we had a lot of log locations using `Info` then they would all need to be revisited w.r.t. to disk filling attacks)

jonatack replied:
Good point. My original thinking was that `info` would be conditional like `debug` and `trace` and that only `warning` and `error` would be unconditional, which would also reduce the volume of default logging. However, that changed during the discussions in https://github.com/bitcoin/bitcoin/pull/25306. Will keep your suggestion in mind while restructuring the update here.

jonatack commented:
Updating and rebasing.

jonatack commented:
Thanks for the detailed feedback, @ajtowns. Will update with it in mind; looks like it will be helpful in simplifying things.
